### PR TITLE
re-usable component for viewer dropdown

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request continues the pattern in #1156 to pull up duplicated logic from plugins and into a re-usable component for viewer select dropdowns.  In doing so, it also gives flexibility for reference vs ID (so that cubeviz will show the more user-friendly "uncert-viewer" names instead of "cubeviz-2" but imviz will still work with IDs since references are never created).  All other functionality should be unchanged - this is mostly just a code refactoring.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by javerbukh** [comment]
Got a traceback when running the MosvizNIRISS example notebook, which links to this section:
https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/mosviz/helper.py#L470-L475

**Comment by kecnry** [comment]
> Got a traceback when running the MosvizNIRISS example notebook

I can't seem to reproduce.  Did you do anything other than running the cells?  (I also doubt that anything in that line should be caused by this PR, but still worth investigating).

I'll work to rebase on main as well since I see there are now conflicts.

**Comment by javerbukh** [comment]
Nothing other than running the cells. Specifically it was line 475 that caused the traceback. I'll try again after you rebase.

File: jdaviz/core/template_mixin.py [hunk 1]
```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
...
+    @property
+    def ref_or_ids(self):
+        return [item['ref_or_id'] for item in self.items]
+
+    @property
+    def selected_item(self):
+        for item in self.items:
+            if item['ref_or_id'] == self.selected:
+                return item
+        # try again but this time allow match to id alone.  Note that _selected_changed
```
Comment (pllim) [inline jdaviz/core/template_mixin.py:512]: Why is this needed? Looks like the tests don't go here according to codecov.

File: jdaviz/core/template_mixin.py [hunk 2]
```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
...
+            else:
+                self._handle_default()
+                raise ValueError(f"{event['new']} not one of {self.ref_or_ids}")
+
+    def _handle_default(self):
+        if self.selected not in self.ref_or_ids:
+            # default to first entry, will trigger any observer on selected
+            self.selected = self.ref_or_ids[0] if len(self.items) else ""
+
+    def _on_viewers_changed(self):
```
Comment (pllim) [inline jdaviz/core/template_mixin.py:541]: This goes with removal of `lambda` above.
```suggestion
    def _on_viewers_changed(self, *args):
```

File: jdaviz/core/template_mixin.py [hunk 3]
```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
...
+        # list of dictionaries with id, ref, ref_or_id
+        def _dict_from_viewer(viewer_item):
+            d = {'id': viewer_item['id']}
+            if viewer_item.get('reference') is not None:
+                d['reference'] = viewer_item['reference']
+                d['ref_or_id'] = viewer_item['reference']
```
Comment (pllim) [inline jdaviz/core/template_mixin.py:547]: This gives me a headache. Haha. We really need to get rid of `reference` some day.

File: jdaviz/core/template_mixin.py [hunk 4]
```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
...
+            else:
+                d['reference'] = None
+                d['ref_or_id'] = viewer_item['id']
+            return d
+
+        self.items = [_dict_from_viewer(self.app._viewer_item_by_id(vid))
```
Comment (pllim) [inline jdaviz/core/template_mixin.py:553]: Why store as a separate `dict` here instead of accessing `app._viewer_store` directly in the mixin?

File: jdaviz/core/template_mixin.py [hunk 5]
```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
...
+        </v-row>
+    """
+    viewer_items = List().tag(sync=True)
+    viewer_selected = Unicode().tag(sync=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.viewer = ViewerSelect(self,
+                                   'viewer_items',
+                                   'viewer_selected')
```
Comment (pllim) [inline jdaviz/core/template_mixin.py:602]:
```suggestion
        self.viewer = ViewerSelect(self, 'viewer_items', 'viewer_selected')
```

**Review by pllim — COMMENTED** [review]
General review right now. I will do some manual testing tomorrow, in case you decide to apply some updates to the code.

In principle, I like the simplification of code across the board. Thanks!

File: jdaviz/core/template_mixin.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/core/template_mixin.py:512]: This is needed in order to support setting the selected item based on the id.  I think its worth keeping, so will add tests to cover it.

File: jdaviz/core/template_mixin.py (hunk 4 above)
Comment (kecnry) [reply jdaviz/core/template_mixin.py:553]: I'm not sure that would work... it needs to be a traitlet so that it can push changes to the UI.

File: jdaviz/core/template_mixin.py [hunk 6]
```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
...
+    """
+    def __init__(self, plugin, items, selected):
+        super().__init__(plugin,
+                         items=items,
+                         selected=selected)
+
+        self.hub.subscribe(self, ViewerAddedMessage,
+                           handler=lambda _: self._on_viewers_changed())
+        self.hub.subscribe(self, ViewerRemovedMessage,
+                           handler=lambda _: self._on_viewers_changed())
```
[the remark this answers is not in the document]
Comment (kecnry) [reply jdaviz/core/template_mixin.py:489]: This was just inherited from the original implementation.  I think sometimes the lambda is warranted, but I'll drop them here in case we ever want to use the message itself.

File: jdaviz/core/template_mixin.py (hunk 2 above)
Comment (kecnry) [reply jdaviz/core/template_mixin.py:541]: I dislike `*args`, but will change this to take `msg`.  The downside to doing this over the lambda is that internal calls will have to pass a message (or we can default to `None`, but then we need a check for `None` if we ever add code that does use the message object).

**Review by javerbukh — APPROVED** [review]
New conda environment fixed the problem I was seeing. After testing all the configurations everything functions as expected. Nice work!

**Review by rosteen — APPROVED** [review]
Works in the contexts where it's currently used, code looks good I think. I'm glad you documented the usage a bit in the mixin docstring, but it does remind me that we're due for a bit of a developer docs refresh at some point. 

Testing this made me notice one small bug, but it's also present on main so I'll report it in a separate issue.
